### PR TITLE
Logic Config

### LINKED ISSUE #8: Replace logic difficulty with individual tech options
E.g. instead of a Normal/Hard SM logic option, have separate toggles for e.g. IBJ, machball, wall jumping, etc. 

Probably a pretty big rewrite.

### PR DESCRIPTION
**Pull request by MattEqualsCoder**
Resolves #8 

@Vivelin This is currently a prototype like last time to make sure you and I are in sync on implementation before I proceed further. Still need to update to allow modifying in the UI, saving them like the other configs, and converting the RandomizerOptions to the LogicConfig. Let me know if you have any concerns with this.

Renamed this to LogicConfig to encompass both tricks and quality of life style logic changes. My thought was the two types would be be in two groups in the UI.

### CODE REVIEW

**Comment by Vivelin** [comment]
I can't help but wonder if there isn't a better way to do this, rather than passing LogicConfig through half the code base. `World` has an instance of the general config which isn't very nice either but makes it easier to access. Maybe something like that?

Otherwise, I never really liked the way the original devs set up the ProgressionExtensions. The Progression class always felt more like an inventory to me. I'd rather take Progression as a parameter to something that replaces the ProgressionExtensions with an actual class.

I'm not really sure, to be honest.

File: src/Randomizer.SMZ3/AdvancedLogic.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
+﻿using Randomizer.Shared;
+using static Randomizer.SMZ3.SMLogic;
+
+namespace Randomizer.SMZ3
+{
+    public class AdvancedLogic
```
Comment (Vivelin) [inline src/Randomizer.SMZ3/AdvancedLogic.cs:6]: One small remark: I'd pull these into an ILogic interface and use that on World instead, so we could swap out different "kinds" of logic more easily later on.

File: src/Randomizer.SMZ3/World.cs [hunk 2]
```diff
@@ -93,6 +94,7 @@ public World(Config config, string player, int id, string guid)
         public IEnumerable<Room> Rooms { get; }
         public IEnumerable<Location> Locations { get; }
         public IEnumerable<Item> Items => Locations.Select(l => l.Item).Where(i => i != null);
+        public AdvancedLogic AdvancedLogic { get; }
```
Comment (Vivelin) [inline src/Randomizer.SMZ3/World.cs:97]: Regarding my other point, I'd also rename the property to simply "Logic".

**Review by Vivelin — APPROVED** [review]

**Comment by Vivelin** [comment]
Oh, one more thing: You could also make the new logic class available on Region directly, which makes all this a little more concise. I know it already has a "Logic" property but considering this more or less supersedes it, I say we just replace it and get rid of the "Hard" logic switch.

**Comment by MattEqualsCoder** [comment]
@Vivelin There are definitely more logic options and tricks that should be added, but this is a start at least. This should be ready now.

File: src/Randomizer.Shared/Models/LogicConfig.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿using System.ComponentModel;
+
+namespace Randomizer.Shared
+{
+    public class LogicConfig
+    {
+        [DisplayName("Prevent Screw Attack Sock Locks"), Description("Requires the morph ball before using screw attacks in locations where you can get stuck"), Category("Logic")]
```
Comment (Vivelin) [inline src/Randomizer.Shared/Models/LogicConfig.cs:7]: Typo in "Sock Lock". I'd change the description to "You're not expected to use Screw Attack without the Morph ball" or something along those lines.

File: src/Randomizer.Shared/Models/LogicConfig.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        public bool PreventScrewAttackSoftLock { get; set; }
+
+        [DisplayName("Require Ten Power Bombs"), Description("Always require at least two power bomb upgrades for navigation in Super Metroid. Will require tracking multiple power bombs."), Category("Logic")]
```
Comment (Vivelin) [inline src/Randomizer.Shared/Models/LogicConfig.cs:10]: I'd change "Always require" to "You're expected to have". Since this would effectively give you another set of power bombs early, maybe change the name to "More power bombs" or something? "Require" doesn't quite feel like the right word to me.

File: src/Randomizer.Shared/Models/LogicConfig.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        public bool RequireTwoPowerBombs { get; set; }
+
+        [DisplayName("Fire Rod for Dark Rooms"), Description("Having the fire rod will open up the back of Hyrule Castle, Eastern Palace Armos Knights, and select rooms in Palace of Darkness"), Category("Tricks")]
+        public bool FireRodDarkRooms { get; set; }
+
+        [DisplayName("Infinite Bomb Jump"), Description("Allow using morph bombs to access high locations with specially timed morph bombs"), Category("Tricks")]
```
Comment (Vivelin) [inline src/Randomizer.Shared/Models/LogicConfig.cs:16]: I'd change this from "Allow using" to "You're expected to use" or "You're expected to be able to use"

File: src/Randomizer.Shared/Models/LogicConfig.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        public bool InfiniteBombJump { get; set; }
+
+        [DisplayName("Parlor Speed Booster Break In"), Description("Allow using the speed booster to burst through to the Terminator Room/Brinstar without morph bombs or screw attack"), Category("Tricks")]
```
Comment (Vivelin) [inline src/Randomizer.Shared/Models/LogicConfig.cs:19]: Same as above

File: src/Randomizer.SMZ3/Logic.cs [hunk 7]
```diff
@@ -1,20 +1,140 @@
...
+        }
+
+        public bool CanParlorSpeedBoost(Progression items)
+        {
+            return items.SpeedBooster && World.Config.LogicConfig.ParlorSpeedBooster;
+        }
+
+        public World World { get; }
+
+        public static List<ItemType[]> GetMissingRequiredItems(Location location, Progression items)
```
Comment (Vivelin) [inline src/Randomizer.SMZ3/Logic.cs:133]: Why the change from IEnumerable to List?

File: src/Randomizer.SMZ3/Logic.cs [hunk 8]
```diff
@@ -1,20 +1,140 @@
...
+                    items.Gravity && items.Morph &&
+                    (World.CanAquire(items, Reward.Agahnim) || items.Hammer && CanLiftLight(items) || CanLiftHeavy(items));
+        }
+
+        public bool CanSafelyUseScrewAttack(Progression items)
+        {
+            return items.ScrewAttack && (!World.Config.LogicConfig.PreventScrewAttackSoftLock || items.Morph);
+        }
+
+        public bool CanFireRodLightDarkRooms(Progression items)
```
Comment (Vivelin) [inline src/Randomizer.SMZ3/Logic.cs:121]: Considering this check is almost always used in conjunction with ` || items.Lamp`, maybe just make that part of this check? Mind you, I'm not too familiar with the fire rod logic thing.

**Review by Vivelin — APPROVED** [review]

File: src/Randomizer.Shared/Models/LogicConfig.cs (hunk 3 above)
Comment (MattEqualsCoder) [reply src/Randomizer.Shared/Models/LogicConfig.cs:7]: What do you have against comfy socks? :(

Updated all of the display names/descriptions

File: src/Randomizer.SMZ3/Logic.cs (hunk 7 above)
Comment (MattEqualsCoder) [reply src/Randomizer.SMZ3/Logic.cs]: Whoops. Was testing something and forgot to back it out.

File: src/Randomizer.SMZ3/Logic.cs (hunk 8 above)
Comment (MattEqualsCoder) [reply src/Randomizer.SMZ3/Logic.cs]: Updated function name to CanPassFireRodDarkRooms and made this change

**Review by Vivelin — APPROVED** [review]
